specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1098 @@
+
+> TODO: Come up with error codes for message failures and how to propagate them to the source. Only the source will be notified.
+
+### Minting and burning native tokens in subnets
+Native tokens are injected to the circulating supply of a subnet by applying top-down messages. These messages lock the amount of tokens included in the `value` of the message in the `SCA` of the parent, and they trigger the minting of new tokens in the subnet when they are executed. To mint new tokens in the subnet, we include a new `SubnetMint` method with `methodNum=5` to the `RewardActor`. `SubnetMint` can only be called by the `SCA` in a subnet through `ApplyMsg()` method when executing a message. `SubnetMint` funds the `SCA` with enough minted native tokens to provide the destination address with its corresponding subnet tokens (see [sample implementation](https://github.com/filecoin-project/eudico/blob/bb52565105f7fe716463b1e09dad7492569089f5/chain/consensus/actors/reward/reward_actor.go#L94)).
+
+Validators in subnets are exclusively rewarded in native-tokens through message fees, so the balance of the `RewardActor` in subnets is only used to increase the circulating supply in a subnet by order of the `SCA`. For new native tokens to be minted in a subnet, the same amount of tokens need to have been locked in the `SCA` of the parent.
+
+> Note: This change of the reward actor is not required for the `rootnet`. We can have a custom bundle with the modified `RewardActor` to be used exclusively for subnets.
+
+On the other hand, bottom-up messages release funds from a subnet (reducing its circulating supply). In this case, burning the funds from the subnet is straightforward. When the bottom-up messages are included in a checkpoint, the `SCA` triggers a message with the `value` amount of native-tokens to the `BurnActorAddress` of the subnet. Once the checkpoint is committed to the parent, and the messages executed, the same amount of native tokens will be released from the parent's `SCA` and sent in a message to the destination addresses of the bottom-up messages.
+
+    
+## Subnet Content Resolution Protocol
+For scalability reasons, when the destination subnet receives a new checkpoint with cross-net messages to be executed, it is only provided with the `CID` of the aggregated messages inside the `CrossMsgMeta`. For the subnet to be able to trigger the corresponding state changes for all the messages, it needs to fetch the payload of messages behind that `CID`, as illustrated in previous sections. The subnet `SCA` where the bottom-up message is generated keeps a `CrossMsgsRegistry` with all `CID`s for `CrossMsgsMeta` propagated (i.e. a content-addressable key-value store), used to fulfill content resolution requests.
```

Comment:
> And this registry will hold on to the messages included in all previous checkpoints, forever, right?

Correct. We keep the latest version of `CrossMsgs`.

> Would IPLD somehow recognise that the b1 store entry is referenced, but b2 is not?

Unfortunately, not. This would need to be handled in the persistence layer (i.e. the datastore used to persist the objects behind the links). This is one of the reasons why I personally think that having a DSL/programming language based on IPLD could be really powerful. The compiler could handle this kind of checks.